File: 0039-combination-sum/0039-combination-sum.cpp
class Solution {
public:
    void backtrack(int start,vector<int>&nums,vector<vector<int>>&result,vector<int>&temp,int sum,int target){
        if(sum==target){
            result.push_back(temp);
            return;
        }
        if(start>=nums.size()||sum>target)return;
        temp.push_back(nums[start]);
        backtrack(start,nums,result,temp,sum+nums[start],target);
        temp.pop_back();
        backtrack(start+1,nums,result,temp,sum,target);
    }
    vector<vector<int>> combinationSum(vector<int>& nums, int target) {
        vector<vector<int>>result;
        vector<int>temp;
        backtrack(0,nums,result,temp,0,target);
        return result;
    }
};

File: 0039-combination-sum/0039-combination-sum.java
class Solution {
    public List<List<Integer>>res=new ArrayList<>();
    public void backtrack(int start,int[]candidates,List<Integer>temp,List<List<Integer>>res,int target,int sum){
        if(sum==target){
            res.add(new ArrayList<>(temp));
            return;
        }
        if(start>=candidates.length||sum>target){
            return;
        }
        temp.add(candidates[start]);
        backtrack(start,candidates,temp,res,target,sum+candidates[start]);
        temp.remove(temp.size()-1);
        backtrack(start+1,candidates,temp,res,target,sum);

    }
    public List<List<Integer>> combinationSum(int[] candidates, int target) {
        
        List<Integer>temp=new ArrayList<>();
        backtrack(0,candidates,temp,res,target,0);
        return res;
    }
}